### DISCUSSION TITLE
Silent error when re-adding a primary key during document addition

### DISCUSSION THREAD

**Discussion by curquiza** [meilisearch/product#115] [Feedback & Feature Proposal]
Currently, we can add the primary key of an index when adding/updating documents.

```ruby
documents = [{ numero: 1, titre: 'Le Rouge et le Noir' }]
index.update_documents(documents, 'numero')
```

The first time the script is run, the index will be created and the primary key will be set to `numero`.

If I re-launch the script, the index is already created, and so is the primary key. But **MeiliSearch does not throw any error despite a primary key is already present and the user might want to be informed about this**. 
This silent error might be problematic if the user tries to send a different primary key compared to the one that is stored in the index: he/she is clearly doing something that he/she should not do.

However, if the primary key that is sent is the same primary key that is already set up, I personally don't like with the fact of raising an error since it brings an unpleasant user experience when coding.
Indeed, users usually send documents by batch, so in a loop, like this:

```
for batch_documents in all_documents:
  index.update_documents([batch_documents], 'numero')
```

Here it will not fail at the first call, but it will at the second one. So the users would need to create an index first with the primary key. This is not convenient: we might not want to make this kind of complexity back (create an index first) knowing that we succeeded to remove it with the implicit index creation handling.

### Conclusion
The silent error is problematic: the users should know what they are doing if they are doing something wrong. However, if the primary key sent is the same as the one already stored, they are not in a dangerous situation when manipulation MeiliSearch.
**A solution would be to raise an error only if the primary key sent is different from the primary key already set in the index**.

_Discussed with @MarinPostma about the problematic silent error.
Suggestion of improvement coming from @bidoubiwa and @eskombro._

**Comment by eskombro** [comment]
100% agreed. From my POV, MeiliSearch should raise an error when the user **tries (voluntarily or not, knowingly or not)** to change the primary key, because this is technically not possible. Other than that, sending the same primary key (as in the loop example) shouldn't raise any error or even warning.

**Comment by curquiza** [comment]
@gmourier this one can interest you

**Reply by gmourier** [reply]
Hey @curquiza,

📡 @irevoire for further information.

>  the users should know what they are doing if they are doing something wrong.

I agree with that statement.

From my understanding, we can't throw a synchronous error but https://github.com/meilisearch/meilisearch/issues/3385 is solving it in an asynchronous manner. (at least, it will not silently accept payload and mark them as succeeded).
